python/ray/experimental/channel/torch_tensor_nccl_channel.py

```diff
@@ -0,0 +1,222 @@
+        return self._reader_registered, "Actor is not a reader"
+
+    def __reduce__(self):
+        return (self.__class__, (self._writer, self._readers, self._typ, self._device))
+
+    def write(
+        self,
+        wrapped: "_TorchTensorWrapper",
+        readers: Optional[List[ray.actor.ActorHandle]] = None,
+    ):
+        value = wrapped.tensor
+        if value.shape != self._typ.shape:
+            raise ValueError(
+                f"torch.Tensor has shape {value.shape}, expected {self._typ.shape}"
+            )
+        if value.dtype != self._typ.dtype:
+            raise ValueError(
+                f"torch.Tensor has shape {value.shape}, expected {self._typ.shape}"
+            )
+        if value.device != self._device:
+            raise ValueError(
+                f"torch.Tensor must be on the default device: {self._device}"
+            )
+
+        if readers is not None:
+            raise NotImplementedError(
+                "TorchTensorNcclChannel.write() not supported for dynamically "
+                "passed readers."
+            )
+
+        # TODO: If there are multiple readers, can replace with a broadcast.
+        for rank in self._reader_ranks:
+            self._nccl_group.send(value, rank)
+
+    def begin_read(self) -> torch.Tensor:
+        buf = torch.zeros(self._typ.shape, dtype=self._typ.dtype, device=self._device)
+        self._nccl_group.recv(buf, self._writer_rank)
+        return buf
+
+    def end_read(self) -> None:
+        return
+
+    def close(self) -> None:
+        self._nccl_group.destroy()
+
+
+def _do_init_nccl_group(self, world_size, comm_id, rank, actor_ids_to_ranks):
+    # assert ray.get_gpu_ids()
+
+    ctx = ChannelContext.get_current()
+    ctx.nccl_group = _NcclGroup(
+        world_size,
+        comm_id,
+        rank,
+        actor_ids_to_ranks,
+        torch.cuda.current_stream().cuda_stream,
+    )
+
+
+@DeveloperAPI
+def _do_check_has_gpu(self) -> bool:
+    return bool(ray.get_gpu_ids())
+
+
+def _init_nccl_group(
+    actors: List[ray.actor.ActorHandle],
+):
+    ctx = ChannelContext.get_current()
+    if ctx.nccl_group is not None:
+        return
+
+    has_gpus = ray.get(
+        [actor.__ray_call__.remote(_do_check_has_gpu) for actor in actors]
+    )
+    for has_gpu, actor in zip(has_gpus, actors):
+        if not has_gpu:
+            raise ValueError(
+                f"Actor {actor} returns a tensor with type hint "
+                'TorchTensor(transport="nccl") but actor does not have a '
+                "GPU assigned by Ray."
+            )
+
+    actor_handles_to_ranks = {actor: rank for rank, actor in enumerate(actors)}
+    actor_ids_to_ranks = {
+        actor._ray_actor_id: rank for actor, rank in actor_handles_to_ranks.items()
+    }
+    assert len(actor_ids_to_ranks) == len(
+        actor_handles_to_ranks
+    ), "actors must be unique"
+
+    from cupy.cuda import nccl
```

Comment:
Does cupy group guaranteed to be isolated?

Reply:
What do you mean?

Reply:
like if the user application uses its own cupy, does it interfere with ours?

Reply:
Do you mean cupy version or a separate NCCL group?

cupy version should be fine, as long as NCCL is supported.

Separate NCCL group can coexist but the user cannot use [concurrent kernels](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/usage/communicators.html#using-multiple-nccl-communicators-concurrently). If the other NCCL group is on the same stream, then it's okay. Otherwise, they need to synchronize with the stream passed here before/after the other NCCL ops. Will add this to the docs.